### PR TITLE
[JSC] Different @@species getter for each constructor

### PR DESCRIPTION
**Pull request by sosukesuzuki**
#### d91d519d4b06ae84e14ccf1cde231683b0b7d6c4
<pre>
[JSC] Different @@species getter for each constructor
<a href="https://bugs.webkit.org/show_bug.cgi?id=267039">https://bugs.webkit.org/show_bug.cgi?id=267039</a>

Reviewed by Keith Miller.

In JavaScriptCore, `Object.getOwnPropertyDescriptor(Array, Symbol.species).get === Object.getOwnPropertyDescriptor(RegExp, Symbol.species).get` evaluates to true.
This patch changes the @@species getters of each constructor to be different objects, so that the result of this expression becomes false.
Remove the global `m_speciesGetterSetter` and create `m_arraySpeciesGetterSetter` and `m_regExpSpeciesGetterSetter` and ... for each constructor.
Instead of passing the `GetterSetter*` type values as arguments to the create function of each constructor, refer to them from the `globalObject`.

<a href="https://tc39.es/ecma262/#sec-get-regexp-@@species">https://tc39.es/ecma262/#sec-get-regexp-@@species</a>
<a href="https://tc39.es/ecma262/#sec-get-array-@@species">https://tc39.es/ecma262/#sec-get-array-@@species</a>
<a href="https://tc39.es/ecma262/#sec-get-%typedarray%-@@species">https://tc39.es/ecma262/#sec-get-%typedarray%-@@species</a>
<a href="https://tc39.es/ecma262/#sec-get-map-@@species">https://tc39.es/ecma262/#sec-get-map-@@species</a>
<a href="https://tc39.es/ecma262/#sec-get-set-@@species">https://tc39.es/ecma262/#sec-get-set-@@species</a>
<a href="https://tc39.es/ecma262/#sec-get-arraybuffer-@@species">https://tc39.es/ecma262/#sec-get-arraybuffer-@@species</a>
<a href="https://tc39.es/ecma262/#sec-sharedarraybuffer-@@species">https://tc39.es/ecma262/#sec-sharedarraybuffer-@@species</a>
<a href="https://tc39.es/ecma262/#sec-get-promise-@@species">https://tc39.es/ecma262/#sec-get-promise-@@species</a>

* JSTests/stress/species-equivalence-typedarray.js: Added.
(shouldBe):
(readTypedArraySymbolSpeciesGetter):
(runTest):
* JSTests/stress/species-equivalence.js: Added.
(shouldNotBe):
(readSymbolSpeciesGetter):
(runTest):
* Source/JavaScriptCore/runtime/ArrayConstructor.cpp:
(JSC::ArrayConstructor::finishCreation):
* Source/JavaScriptCore/runtime/ArrayConstructor.h:
* Source/JavaScriptCore/runtime/BigIntConstructor.h:
* Source/JavaScriptCore/runtime/BooleanConstructor.cpp:
(JSC::BooleanConstructor::create):
* Source/JavaScriptCore/runtime/BooleanConstructor.h:
* Source/JavaScriptCore/runtime/DateConstructor.h:
* Source/JavaScriptCore/runtime/ErrorConstructor.h:
* Source/JavaScriptCore/runtime/FinalizationRegistryConstructor.h:
* Source/JavaScriptCore/runtime/JSArrayBufferConstructor.cpp:
(JSC::JSGenericArrayBufferConstructor&lt;sharingMode&gt;::finishCreation):
* Source/JavaScriptCore/runtime/JSArrayBufferConstructor.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
(JSC::JSGlobalObject::tryInstallSpeciesWatchpoint):
(JSC::JSGlobalObject::installArraySpeciesWatchpoint):
(JSC::JSGlobalObject::tryInstallArrayBufferSpeciesWatchpoint):
(JSC::JSGlobalObject::tryInstallTypedArraySpeciesWatchpoint):
(JSC::JSGlobalObject::installTypedArrayConstructorSpeciesWatchpoint):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::arraySpeciesGetterSetter const):
(JSC::JSGlobalObject::typedArraySpeciesGetterSetter const):
(JSC::JSGlobalObject::speciesGetterSetter const): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::JSGlobalObject::arrayBufferSpeciesGetterSetter const):
* Source/JavaScriptCore/runtime/JSInternalPromiseConstructor.cpp:
(JSC::JSInternalPromiseConstructor::create):
* Source/JavaScriptCore/runtime/JSInternalPromiseConstructor.h:
* Source/JavaScriptCore/runtime/JSPromiseConstructor.cpp:
(JSC::JSPromiseConstructor::create):
(JSC::JSPromiseConstructor::finishCreation):
* Source/JavaScriptCore/runtime/JSPromiseConstructor.h:
* Source/JavaScriptCore/runtime/JSTypedArrayViewConstructor.cpp:
(JSC::JSTypedArrayViewConstructor::finishCreation):
* Source/JavaScriptCore/runtime/JSTypedArrayViewConstructor.h:
* Source/JavaScriptCore/runtime/MapConstructor.cpp:
(JSC::MapConstructor::finishCreation):
* Source/JavaScriptCore/runtime/MapConstructor.h:
* Source/JavaScriptCore/runtime/NumberConstructor.cpp:
(JSC::NumberConstructor::create):
* Source/JavaScriptCore/runtime/NumberConstructor.h:
* Source/JavaScriptCore/runtime/RegExpConstructor.cpp:
(JSC::RegExpConstructor::finishCreation):
* Source/JavaScriptCore/runtime/RegExpConstructor.h:
* Source/JavaScriptCore/runtime/SetConstructor.cpp:
(JSC::SetConstructor::finishCreation):
* Source/JavaScriptCore/runtime/SetConstructor.h:
* Source/JavaScriptCore/runtime/ShadowRealmConstructor.h:
* Source/JavaScriptCore/runtime/StringConstructor.cpp:
(JSC::StringConstructor::create):
* Source/JavaScriptCore/runtime/StringConstructor.h:
* Source/JavaScriptCore/runtime/SymbolConstructor.h:
* Source/JavaScriptCore/runtime/WeakMapConstructor.h:
* Source/JavaScriptCore/runtime/WeakObjectRefConstructor.h:
* Source/JavaScriptCore/runtime/WeakSetConstructor.h:

Canonical link: <a href="https://commits.webkit.org/275064@main">https://commits.webkit.org/275064@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/06a33afe5b8ad64adb9587a0895c11f88e1341d2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40757 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19770 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43135 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43315 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36847 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22755 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17101 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33801 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41331 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16716 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35125 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14405 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14495 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36114 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44587 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/34210 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36971 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36427 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40170 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/40383 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15572 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12775 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38511 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17191 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/47393 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9145 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17242 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9736 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16835 "Built successfully") | | |

### CODE REVIEW

File: Source/JavaScriptCore/runtime/JSGlobalObject.h
Comment (kmiller68) [inline Source/JavaScriptCore/runtime/JSGlobalObject.h]: I think we don't need these saved on the global object as we don't watchpoint them so we should save a little memory. Can you construct these in the respective finishCreation instead?

**Review by kmiller68 — CHANGES_REQUESTED** [review]
Thanks for the patch! I have one comment.

File: Source/JavaScriptCore/runtime/JSGlobalObject.h
Comment (sosukesuzuki) [reply Source/JavaScriptCore/runtime/JSGlobalObject.h]: Thank you for your review! Like this?

```diff
void RegExpConstructor::finishCreation(VM& vm, RegExpPrototype* regExpPrototype)
{
    Base::finishCreation(vm, 2, vm.propertyNames->RegExp.string(), PropertyAdditionMode::WithoutStructureTransition);
    ASSERT(inherits(info()));

    putDirectWithoutTransition(vm, vm.propertyNames->prototype, regExpPrototype, PropertyAttribute::DontEnum | PropertyAttribute::DontDelete | PropertyAttribute::ReadOnly);
-   putDirectNonIndexAccessorWithoutTransition(vm, vm.propertyNames->speciesSymbol, globalObject->regExpSpeciesGetterSetter(), PropertyAttribute::Accessor | PropertyAttribute::ReadOnly | PropertyAttribute::DontEnum);
+   GetterSetter* speciesGetterSetter = GetterSetter::create(vm, globalObject, JSFunction::create(vm, globalObject, 0, "get [Symbol.species]"_s, globalFuncSpeciesGetter, ImplementationVisibility::Public, SpeciesGetterIntrinsic), nullptr);
+   putDirectNonIndexAccessorWithoutTransition(vm, vm.propertyNames->speciesSymbol, speciesGetterSetter, PropertyAttribute::Accessor | PropertyAttribute::ReadOnly | PropertyAttribute::DontEnum)
;
```

Comment (sosukesuzuki) [reply Source/JavaScriptCore/runtime/JSGlobalObject.h]: We need to pass an instance of `speciesGetterSetter` as an argument to `tryInstallSpeciesWatchpoint`. If we create them in `finishCreation`, how can we get that instance from the global object side?

https://github.com/WebKit/WebKit/blob/9c99737fd2d825b16daf7e8931842b4d2a2c4448/Source/JavaScriptCore/runtime/JSGlobalObject.cpp#L2770

**Comment by hyjorc1** [comment]
From the commit message:
> This patch change the @@species getters of each constructor to be different objects ...

change -> changes

File: Source/JavaScriptCore/runtime/JSGlobalObject.h
Comment (kmiller68) [reply Source/JavaScriptCore/runtime/JSGlobalObject.h]: That code looks right.

Just for the ones that aren't passed to `tryInstallSpeciesWatchpoint`, namely for RegExp, Map, Set, and Promise.

Comment (sosukesuzuki) [reply Source/JavaScriptCore/runtime/JSGlobalObject.h]: Thank you! Updated

File: Source/JavaScriptCore/runtime/JSGlobalObject.h
Comment (kmiller68) [inline Source/JavaScriptCore/runtime/JSGlobalObject.h]: Should be: `inline GetterSetter* arrayBufferSpeciesGetterSetter(ArrayBufferSharingMode) const;`

**Review by kmiller68 — CHANGES_REQUESTED** [review]
One minor change otherwise LGTM.

**Review by kmiller68 — APPROVED** [review]
r=me.

**Comment by kmiller68** [comment]
Ping me if you when you're ready for me to commit queue this.

**Comment by sosukesuzuki** [comment]
@kmiller68 Thank you for letting me know. I'm ready on my end, but we might need to wait for the CI to complete? This is my first PR to WebKit so I'm not familiar with the process yet.

**Comment by kmiller68** [comment]
No worries. `safe-merge-queue` label will wait for CI to finish before merging. Thanks again for the patch!

**Pull request merged**
